File: src/main/java/com/gitlab/lae/intellij/actions/tree/ui/ActionPresentation.java
package com.gitlab.lae.intellij.actions.tree.ui;

import com.google.auto.value.AutoValue;
import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.Presentation;

import javax.annotation.Nullable;
import javax.swing.*;
import java.awt.event.ActionEvent;
import java.util.List;
import java.util.function.BiConsumer;

@AutoValue
public abstract class ActionPresentation {
    ActionPresentation() {
    }

    public static ActionPresentation create(
            Presentation presentation,
            List<KeyStroke> keys,
            String separatorAbove,
            Boolean sticky,
            AnAction action
    ) {
        return new AutoValue_ActionPresentation(
                presentation,
                keys,
                separatorAbove,
                sticky,
                action
        );
    }

    public abstract Presentation presentation();

    public abstract List<KeyStroke> keys();

    @Nullable
    abstract String separatorAbove();

    public abstract Boolean sticky();

    public abstract AnAction action();

    @Override
    public String toString() {
        String text = presentation().getText();
        if (text == null) {
            text = "";
        }
        return text;
    }

    public void registerShortcuts(
            JList<?> list,
            BiConsumer<ActionPresentation, ActionEvent> consumer
    ) {
        if (keys().isEmpty()) {
            return;
        }
        InputMap inputMap = list.getInputMap();
        ActionMap actionMap = list.getActionMap();
        for (KeyStroke key : keys()) {
            inputMap.put(key, key);
            actionMap.put(key, new AbstractAction() {
                @Override
                public void actionPerformed(ActionEvent e) {
                    consumer.accept(ActionPresentation.this, e);
                }
            });
        }
    }
}

File: src/main/kotlin/com/gitlab/lae/intellij/actions/tree/ui/ActionPresentation.kt
package com.gitlab.lae.intellij.actions.tree.ui

import com.gitlab.lae.intellij.actions.tree.ActionNode.Companion.ACTION_PLACE
import com.gitlab.lae.intellij.actions.tree.util.setEnabledModalContext
import com.intellij.openapi.actionSystem.ActionManager
import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.DataContext
import com.intellij.openapi.actionSystem.Presentation
import com.intellij.openapi.actionSystem.UpdateSession
import com.intellij.openapi.actionSystem.ex.ActionUtil.performDumbAwareUpdate
import java.awt.event.ActionEvent
import javax.swing.AbstractAction
import javax.swing.JList
import javax.swing.KeyStroke

data class ActionPresentation(
  val presentation: Presentation,
  val keys: List<KeyStroke>,
  val separatorAbove: String?,
  val sticky: Boolean,
  val action: AnAction,
) {

  override fun toString() = presentation.text ?: ""

  fun registerShortcuts(
    list: JList<*>,
    consumer: (ActionPresentation, ActionEvent) -> Unit,
  ) {
    if (keys.isEmpty()) {
      return
    }
    val inputMap = list.inputMap
    val actionMap = list.actionMap
    for (key in keys) {
      inputMap.put(key, key)
      actionMap.put(
        key,
        object : AbstractAction() {
          override fun actionPerformed(e: ActionEvent) {
            consumer(this@ActionPresentation, e)
          }
        },
      )
    }
  }

  fun update(
    actionManager: ActionManager,
    dataContext: DataContext,
  ) {
    val event = AnActionEvent(
      null,
      dataContext,
      ACTION_PLACE,
      presentation,
      actionManager,
      0,
    )
    event.setInjectedContext(action.isInInjectedContext)

    // Check for empty session to avoid UnsupportedOperationException
    if (event.updateSession != UpdateSession.EMPTY) {
      performDumbAwareUpdate(action, event, false)
    }

    setEnabledModalContext(event, action)
  }

  companion object {
    fun create(
      action: AnAction,
      keys: List<KeyStroke>,
      separatorAbove: String?,
      sticky: Boolean,
    ) = ActionPresentation(
      action.templatePresentation.clone(),
      keys,
      separatorAbove,
      sticky,
      action,
    )
  }
}
